src/mysql_conn.erl

```diff
@@ -91,54 +92,109 @@ init(Opts) ->
 
     %% If buffer wasn't specifically defined make it at least as
     %% large as recbuf, as suggested by the inet:setopts() docs.
     case proplists:is_defined(buffer, TcpOpts) of
         true ->
             ok;
         false ->
-            {ok, [{buffer, Buffer}]} = inet:getopts(Socket0, [buffer]),
-            {ok, [{recbuf, Recbuf}]} = inet:getopts(Socket0, [recbuf]),
-            ok = inet:setopts(Socket0,[{buffer, max(Buffer, Recbuf)}])
+            {ok, [{buffer, Buffer}]} = inet:getopts(Socket, [buffer]),
+            {ok, [{recbuf, Recbuf}]} = inet:getopts(Socket, [recbuf]),
+            ok = inet:setopts(Socket, [{buffer, max(Buffer, Recbuf)}])
     end,
 
+    do_connect2(State#state{socket = Socket}).
+
+do_connect2(#state{socket = Socket0, ssl_opts = SSLOpts,
+        user = User, password = Password, database = Database,
+        cap_found_rows = SetFoundRows} = State0) ->
     %% Exchange handshake communication.
-    Result = mysql_protocol:handshake(User, Password, Database, SockMod0, SSLOpts,
+    Result = mysql_protocol:handshake(User, Password, Database, gen_tcp, SSLOpts,
                                       Socket0, SetFoundRows),
     case Result of
         {ok, Handshake, SockMod, Socket} ->
             setopts(SockMod, Socket, [{active, once}]),
             #handshake{server_version = Version, connection_id = ConnId,
                        status = Status,
                        auth_plugin_data = AuthPluginData} = Handshake,
-            State = #state{server_version = Version, connection_id = ConnId,
+            State1 = State0#state{server_version = Version, connection_id = ConnId,
                            sockmod = SockMod,
                            socket = Socket,
-                           ssl_opts = SSLOpts,
-                           host = Host, port = Port,
-                           user = User, password = Password,
                            auth_plugin_data = AuthPluginData,
-                           status = Status,
-                           log_warnings = LogWarn,
-                           ping_timeout = PingTimeout,
-                           query_timeout = Timeout,
-                           query_cache_time = QueryCacheTime,
-                           cap_found_rows = (SetFoundRows =:= true)},
-            case execute_on_connect(Queries, Prepares, State) of
-                {ok, State1} ->
-                    process_flag(trap_exit, true),
-                    State2 = schedule_ping(State1),
-                    {ok, State2};
-                {error, Reason} ->
-                    {stop, Reason}
-            end;
+                           status = Status},
+            do_post_connect(State1);
         #error{} = E ->
-            {stop, error_to_reason(E)}
+            {error, error_to_reason(E)}
+    end.
+
+do_post_connect(#state{queries = Queries, prepares = Prepares} = State) ->
+    case execute_on_connect(Queries, Prepares, State) of
```

Comment:
To not add to technical dept, this can be used in `change_user` as well. It has almost the same logic.

Reply:
Hm, I don't know if it achieves much (unless I missed the point), but did as suggested.

Reply:
Looking back at it all, I get the feeling that `mysql_conn` has overall become a bit messy over time. When this one is through, I may look into pulling some things more straight, I guess :)

Reply:
That would be nice!